server/core_tournament.go

```diff
@@ -453,49 +453,58 @@ func TournamentRecordWrite(ctx context.Context, logger *zap.Logger, db *sql.DB,
 			return nil, err
 		}
 
 		if err := ExecuteInTx(ctx, tx, func() error {
-			if err := tx.QueryRowContext(ctx, query, params...).Scan(&dbNumScore, &dbMaxNumScore); err != nil {
+			recordQueryResult, err := tx.ExecContext(ctx, query, params...)
+			if err != nil {
 				return err
 			}
 
-			// Check if the max number of submissions has been reached.
-			if dbMaxNumScore > 0 && dbNumScore > dbMaxNumScore {
-				return ErrTournamentWriteMaxNumScoreReached
-			}
+			// A record was inserted or updated
+			if rowsAffected, _ := recordQueryResult.RowsAffected(); rowsAffected > 0 {
+				var dbNumScore int
+				var dbMaxNumScore int
 
-			// Check if we need to increment the tournament score count by checking if this was a newly inserted record.
-			if dbNumScore <= 1 {
-				res, err := tx.ExecContext(ctx, "UPDATE leaderboard SET size = size + 1 WHERE id = $1 AND (max_size = 0 OR size < max_size)", leaderboard.Id)
+				err := tx.QueryRowContext(ctx, "SELECT num_score, max_num_score FROM leaderboard_record WHERE leaderboard_id = $1 AND owner_id = $2 AND expiry_time = $3", leaderboard.Id, ownerId, expiryTime).Scan(&dbNumScore, &dbMaxNumScore)
 				if err != nil {
-					logger.Error("Error updating tournament size", zap.Error(err))
+					logger.Error("Error reading leaderboard record.", zap.Error(err))
 					return err
 				}
-				if rowsAffected, _ := res.RowsAffected(); rowsAffected != 1 {
-					// If the update failed then the tournament had a max size and it was met or exceeded.
-					return ErrTournamentMaxSizeReached
+
+				// Check if the max number of submissions has been reached.
+				if dbMaxNumScore > 0 && dbNumScore > dbMaxNumScore {
+					return ErrTournamentWriteMaxNumScoreReached
+				}
+
+				// Check if we need to increment the tournament score count by checking if this was a newly inserted record.
+				if dbNumScore <= 1 {
+					res, err := tx.ExecContext(ctx, "UPDATE leaderboard SET size = size + 1 WHERE id = $1 AND (max_size = 0 OR size < max_size)", leaderboard.Id)
+					if err != nil {
+						logger.Error("Error updating tournament size", zap.Error(err))
+						return err
+					}
+					if rowsAffected, _ := res.RowsAffected(); rowsAffected != 1 {
+						// If the update failed then the tournament had a max size and it was met or exceeded.
+						return ErrTournamentMaxSizeReached
+					}
 				}
 			}
 
 			return nil
 		}); err != nil {
 			if err == ErrTournamentWriteMaxNumScoreReached || err == ErrTournamentMaxSizeReached {
 				logger.Info("Aborted writing tournament record", zap.String("reason", err.Error()), zap.String("tournament_id", tournamentId), zap.String("owner_id", ownerId.String()))
-			} else {
-				logger.Error("Could not write tournament record", zap.Error(err), zap.String("tournament_id", tournamentId), zap.String("owner_id", ownerId.String()))
 			}
+
+			logger.Error("Could not write tournament record", zap.Error(err), zap.String("tournament_id", tournamentId), zap.String("owner_id", ownerId.String()))
```

Comment:
I think the `else {` clause here should be kept. If we don't keep it then an error of `ErrTournamentMaxSizeReached` will print both the info and the error message and return early.